### DISCUSSION TITLE
Middleware questions

### DISCUSSION THREAD

**Discussion by tradical-allstripes** [prisma/prisma#5123] [Q&A]
Hey all!

Just wanted a little clarification on the following things:

1. My understanding is that the ```next``` function argument is what runs the actual query after all the "before" middleware runs. As such, the logic *before* the call to ```next``` runs before the query, and everything *after* ```next``` runs after the query. Can you confirm this?
1. Is the return value from the final ```next``` call the data model object?
1. What precisely does the return value of the middleware do? Is it passed to the next middleware if not the last, and, if it is the last, returns the final data model object from the query?
1. It looks like the "after" logic runs in reverse order of definition. Please confirm.

I think these things are pretty fair to assume after reading the documentation closely and making inferences, but I've run into cases in the past where even a fair assumption I made caused pretty real problems, so I'd love to see these things clarified in the documentation. For example, the part that says:

> If you have multiple middlewares, the running order for each separate query is:
>
> * All logic before await next(params)
> * All logic after await next(params)

Could say (if my assumptions are correct):

> Here is the running order for middleware logic:
>
> * All logic before ```await next(params)``` in all assigned middleware, in the order of definition
> * The query
> * All logic after ```await next(params)``` in all assigned middleware, in *reverse order of definition*
(Or you could provide a note that specifies the part about reverse order on the after middleware.)

And then there should be a part of the params section that gives API reference for the ```next``` parameter.

**Comment by ryands17** [comment]
Hey @tradical-allstripes 👋 

1. The actual query will be executed in the last middleware. So as specified in the example [here](https://www.prisma.io/docs/concepts/components/prisma-client/middleware#running-order) the call to the db to fetch data will be in the second middleware (the one where `2` and `3` are logged).
2. Yes that will be the data from the db.
3. Yeah the final middleware will return the data and the previous middleware can manipulate it by either changing the structure or something like that.
4. Yes the `after` logic for the middlewares are run in reverse order and think of it as the final middleware returning data to the one above it forming a chain.

>  but I've run into cases in the past where even a fair assumption I made caused pretty real problems

Could you explain a bit about what error you faced as the order specified in the docs is actually the correct order.

> And then there should be a part of the params section that gives API reference for the next parameter.

As per the type definitions, this is what the next function accepts:
![image](https://user-images.githubusercontent.com/19697099/104919810-db208f80-59bc-11eb-980c-b162eff3e8ff.png)

Let me know if that answers your question :)

**Comment by tradical-allstripes** [comment]
Thanks for the reply! That is very helpful, yes.

> Could you explain a bit about what error you faced as the order specified in the docs is actually the correct order.

It wasn't in this case where I had a problem; I inferred from the code example that the correct order was reverse of definition and that worked out fine. It's just in the past I know I've made false assumptions about the way a particular framework functioned based on their examples when it wasn't specified explicitly in the documentation. So really all I'm suggesting is that the documentation be just a little more clear about that specific bit, instead of relying on users to infer from the code example that this is the case. I'm just wary of inferring things from examples and reading code, as those things often miss context or are not as clear as explicit documentation.

> As per the type definitions, this is what the next function accepts:

Thanks for the info! I think it would be good to make note of this in the documentation as well.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
